test/e2e/framework/conformance/conformance.go

```diff
@@ -0,0 +1,845 @@
+	//
+	// For cluster-scoped resources the test namespace can be used
+	// to create a name which does not conflict with other objects
+	// because it is unique while the test runs.
+	//
+	// It does not need to be set for namespaced resources because the
+	// caller will ensure that. The caller cannot do that for the name because
+	// different resources have different rules for what names are valid
+	GetInitialObject(namespace string) *unstructured.Unstructured
+
+	// GetUpdateSpec modifies an existing object.
+	// It gets called for the result of creating the initial object.
+	//
+	// Ideally it should  change the spec (hence the name).
+	// If that is impossible, then adding some label is also okay.
+	// The goal is to add some fields that can be checked for
+	// after an Update.
+	//
+	// May modify and return the input object.
+	GetUpdateSpec(object *unstructured.Unstructured) *unstructured.Unstructured
+
+	// GetUpdateStatus modifies the status of an existing object.
+	// It gets called for the result of creating the initial object
+	// and then updating its spec.
+	//
+	// May modify and return the input object.
+	GetUpdateStatus(object *unstructured.Unstructured) *unstructured.Unstructured
+
+	// GetPatchSpec describes how to generate patches.
+	//
+	// Each patch is applied to the initial object by itself, without the other patches.
+	// An empty slice is valid and disables testing of patching. This may not be sufficient
+	// for full conformance testing of the resource.
+	//
+	// If content verification is enabled, then this must cause the same change as GetUpdateSpec
+	// because verification of the patch result uses the GetUpdateSpec result as reference.
+	GetPatchSpec() []Patch
+
+	// GetPatchStatus is like GetPatchSpec for the status.
+	//
+	// The initial object with the updated spec gets patched,
+	// so the result must match the result of GetUpdateStatus
+	// applied to GetInitialObject if content verification is
+	// enabled.
+	GetPatchStatus(object *unstructured.Unstructured) []Patch
+}
+
+// Patch contains the parameters for a Patch API call.
+//
+// The data must match an existing object.
+//
+// There's no retry loop because of conflicts, so the patch should not include
+// a check of the ResourceVersion. Checking the UID in the patch is encouraged to prevent
+// patching a replaced resource.
+type Patch struct {
+	GetData func(object *unstructured.Unstructured) []byte
+	Type    types.PatchType
+}
+
+// ResourceTestcase provides test data for testing operations for a resource.
+// Test data is based on the native Go type of the resource.
+//
+// The data is used like this:
+// - create InitialSpec -> update with UpdateSpec -> update status with UpdateStatus
+// - create InitialSpec -> apply StrategicMergePatchSpec and compare against UpdateSpec -> apply StrategicMergePatchStatus and compare against UpdateStatus
+type ResourceTestcase[T any] struct {
```

Comment:
not sure if it makes it easier or harder to work with, but I _think_ you could constrain T to be an implementation of runtime.Object, and then deal with `T` throughout instead of `*T`

```suggestion
type ResourceTestcase[T runtime.Object] struct {
```

(non-blocking, feel free to do it if it's easy and works well)